rfcs/cross_context_messaging.md

```diff
@@ -0,0 +1,122 @@
+solved by the COOP tests with
+[dispatcher.js](https://github.com/web-platform-tests/wpt/blob/7b0ebaccc62b566a1965396e5be7bb2bc06f841f/html/cross-origin-opener-policy/resources/dispatcher.js). This
+provides functions `send(uuid, message)` and `receive(uuid,
+maybe_timeout)` which are used to send and recieve messages between
+contexts given a shared key `uuid`. Typically the way this works is a
+specific resource load is given a `uuid` via a query parameter in the
+url. That resource will then `recieve()` messages using that `uuid`
+whereas other contexts, such as that containing the main test, can
+send messages using `uuid`.
+
+The backend transport layer is the wptserve stash, which stores a list
+of messages per `uuid`. A Python script `pop()`s the message from the
+from of the queue when it handles a `GET` request with `uuid` as a
+query parameter, and pushes the request body to the stash when it
+handles a `POST` request, also with a `uuid` parameter.
+
+This functionality is useful enough that it was adopted for Chrome's
+bfcache tests, which suggests it's likely to be useful as a primitive
+in many cases where tests involve multiple contexts or navigations.
+
+It is proposed that we lift this functionality into testdriver itself,
+integrating it with the context system used to target. The underlying
+communication is poll-based, so as a generalisation we expose three
+functions on `test_driver`, each of which returns a promise that is
+resolved when the relevant action is complete:
+
+```
+test_driver.send(message, context);
+test_driver.poll();
+test_driver.recv(timeout=null)
+```
+
+`test_driver.send(mesage, context)` Sends `message` to the context
+`context`. `message` must be JSON-encodable. Unlike the existing
+implementation we don't send `message` directly, but instead wrap it
+in a js object like:
+
+```
+{
+  src: <uuid of sending context>,
+  data: message,
+}
+```
+Including `src` in the wrapper means we're always able to reply to a
+message without needing to explictly communicate the uuid for the
+source context.
+
+`test_driver.poll()` polls the server for a message with the `uuid`
+for the current context. The return value is `null` if there is no
+message, or an object representing the wrapped message, if there is.
+
+`test_driver.recv(timeout=null)` waits at most `timeout` ms to receive
+a message, and throws an error if no message is received. If the
+timeout is null, the function polls indefinitely.
+
+As with the existing implementation, the backend is based on the
+server stash. Since this is standard wpt infrastructure not tied to
+wptrunner, it's possible to provide an implementation directly in
+`testdriver.js` as part of `test_driver_internal`, meaning that it can
+be overridden with a different transport if required, but there is no
+downstream work required for implementors as part of this feature.
+
+The use of the stash does present some promises; to avoid races if
```

Comment:
Promises? Or challenges?

Reply:
It promises some challenges :p (but yes, good catch).